dropshot/src/handler.rs

```diff
@@ -197,6 +201,91 @@ pub trait Extractor: Send + Sync + Sized {
     ) -> ExtractorMetadata;
 }
 
+/**
+ * `WebSocketExt` defines a function for upgrading a RequestContext to a websocket
+ * connection. This automatically handles checking the original request's headers
+ * to make sure it is a valid websocket connection.
+ */
+#[async_trait]
+pub trait WebSocketExt {
+    /**
+     * Upgrade a request to a websocket connection.
+     */
+    async fn upgrade<F, U>(
+        &self,
+        func: F,
+    ) -> Result<HttpResponseUpgraded, HttpError>
+    where
+        F: FnOnce(crate::websocket::WebSocket) -> U + Send + 'static,
+        U: Future<Output = ()> + Send + 'static;
+}
+
+#[async_trait]
+impl<Context: ServerContext> WebSocketExt for RequestContext<Context> {
+    async fn upgrade<F, U>(
+        &self,
+        func: F,
+    ) -> Result<HttpResponseUpgraded, HttpError>
+    where
+        F: FnOnce(crate::websocket::WebSocket) -> U + Send + 'static,
+        U: Future<Output = ()> + Send + 'static,
+    {
+        // We want to parse the headers from the request context.
+        let mut request = self.request.lock().await;
+
+        // We want to make sure we have the websocket upgrade header.
+        let h: Option<headers::Connection> = request.headers().typed_get();
+        if let Some(h) = h {
+            if !h.contains(&http::header::UPGRADE) {
+                return Err(HttpError::for_bad_request(
+                    None,
+                    "Connection header did not include 'upgrade'".to_string(),
+                ));
+            }
+        } else {
+            return Err(HttpError::for_bad_request(
+                None,
+                "Connection header not sent".to_string(),
+            ));
+        }
+
+        // Now get the secure websocket key.
+        let h: Option<headers::SecWebsocketKey> = request.headers().typed_get();
+        let websocket_key = if let Some(h) = h {
+            h
+        } else {
+            return Err(HttpError::for_bad_request(
+                None,
+                "Websocket key not sent".to_string(),
+            ));
+        };
+
+        // Spawn a task to handle the websocket connection.
+        // Use the hyper feature of upgrading a connection.
+        let fut = hyper::upgrade::on(request.deref_mut())
```

Comment:
```suggestion
        let fut = hyper::upgrade::on(&mut *request)
```

Don't need to explicitly import and call derefmut.